### ISSUE TITLE
[silent BUG] Saving/Restoring/Seeding PPO model when action_spec has multiple actions

### ISSUE THREAD

**Issue by unrealwill** [reinforceio/tensorforce#453]
I'm still a novice with tensorforce.
I'm trying to save my ppo agent after training. 
The agent train well but when I save the model, stop the program, then relaunch the program, restore model then the agent performance is as from scratch, whereas it was working well before.

To save/restore I use : 
```
agent.save_model(directory=directory)
agent.restore_model(directory=directory)
```

I have looked using : 
```
 tf_weights =agent.model.network.get_variables()
 weights = agent.model.session.run(tf_weights)
 print(weights)
```
That the saved weights are correctly restored.

I tried to set some seed using
` tf.set_random_seed(42) `
at the beginning of my program
in the hope to obtain reproducible results (my env is fully deterministic), but upon two sequential launch from the same restored weight, I get different actions for the same input. 

First run first action after restore : 
```
input : 
[[ 0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.
   0.05  0.    0.    0.    0.05  0.    0.    0.    0.05]]
action : 
[-1.65043855 -0.12582253  0.33019719 -0.42400551  0.39128172 -0.1892394
 -1.38783872 -0.84797424 -0.76125687 -0.44233581  0.2647087   0.57517719]
```
Second run first action after restore : 
```
input : 
[[ 0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.
   0.05  0.    0.    0.    0.05  0.    0.    0.    0.05]]
action : 
[ 0.00452828  1.70186901  0.18290332  0.1153125   0.80178595 -1.31738091
  0.2404308  -0.16986398 -1.69459999  2.09507513 -0.46165684 -0.34024456]
```

I have disabled exploration and created the agent with : 
```
layerSize=300
 actions = {}
 for i in range(12):
       actions[str(i)] = {'type': 'float'}
network_spec = [
            dict(type='dense', size=layerSize, activation='selu'),
            dict(type='dense', size=layerSize, activation='selu'),
            dict(type='dense', size=layerSize, activation='selu')
        ]
agent = PPOAgent(
            states=dict(type='float', shape=(12+9,)),
            actions=actions,
            batching_capacity=1000,
            network=network_spec,
            states_preprocessing=None,
            entropy_regularization=1e-3,
            actions_exploration=None,
            step_optimizer=dict(
                type='adam',
                learning_rate=1e-5
            ),
        )
```

Are there some extra parameters which needs to be saved when saving a PPO agent ? (maybe the parameters of the last layer which are used to generate the mean and variance of the gaussians needed to generate the continuous action).

```
tensorforce.__version__
'0.4.2'
```

Thanks

**Comment by unrealwill** [comment]
If instead of defining the action as 12 continuous actions, I define it as one action of shape 12,
`actions=dict(type='float', shape=(12,) )`

This mitigation will work for now for me, but it confirms that there is an issue when restoring multiple action.

The bug disappear i.e. : 
First Run First action after restore : 
```
input : 
[[ 0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.
   0.05  0.    0.    0.    0.05  0.    0.    0.    0.05]]
action : 
[-0.09666101  0.10412237  0.06040742 -0.07674448  0.16232589  0.19782126
  0.07835855 -0.09264671 -0.14772451  0.194968    0.2040552   0.21041678]
```
Second run First action after restore : 
```
input : 
[[ 0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.
   0.05  0.    0.    0.    0.05  0.    0.    0.    0.05]]
action : 
[-0.09666101  0.10412237  0.06040742 -0.07674448  0.16232589  0.19782126
  0.07835855 -0.09264671 -0.14772451  0.194968    0.2040552   0.21041678]
```

**Comment by AlexKuhnle** [comment]
Shouldn't you get a dict of output actions, if you specify a dict of actions?

**Comment by unrealwill** [comment]
@AlexKuhnle Yes I have a dict of actions which I convert back using : 
```
for i in range(12):
      action[i] = actiondict[str(i)][0]
```
I

**Comment by AlexKuhnle** [comment]
I tried reproducing the problem with `Reverse-v0`, which also has [multiple actions](https://github.com/openai/gym/wiki/Table-of-environments), but it seems to work alright. Could you try as well, and/or share a full minimal code example?

**Comment by unrealwill** [comment]
I have the bug in this version : 
https://github.com/GistNoesis/OpenDogSimulator/blob/a8406cd1dafc176c675039cf705551c95415d1e4/trainEnvWithPPO.py

not very minimal but almost (there is plenty of dead code).
I'll try to make a minimal one. (probably won't even need a gym environment)

**Comment by unrealwill** [comment]
```
from tensorforce.agents import PPOAgent
import numpy as np

def demo():
    actions = {}
    for i in range(12):
        actions[str(i)] = {'type': 'float'}  # 'num_actions': 10
    layerSize = 300
    network_spec = [
        dict(type='dense', size=layerSize, activation='tanh'),
        dict(type='dense', size=layerSize, activation='tanh')
    ]

    states = {"jp": dict(type='float', shape=(10,))}

    agent = PPOAgent(
        states=states,
        actions=actions,
        batching_capacity=1000,
        network=network_spec,
        step_optimizer=dict(
            type='adam',
            learning_rate=1e-5
        ),
    )

    try :
        agent.restore_model("./bug/")
    except:
        agent.save_model("./bug/")

    _states = {}
    _states["jp"] = np.arange(10)
    print( agent.act( _states, deterministic=True ) )



if __name__ == "__main__":
    demo()
```
On first run : 
`{'7': -0.84931278, '5': -0.61228681, '9': 0.91923809, '6': -1.2751533, '2': -1.3868549, '4': 0.449507, '0': -0.73226953, '8': 0.50917882, '11': 0.45640573, '1': 0.11466928, '3': -0.23303002, '10': -0.77098644}`
On second run : 
`{'5': -0.84931278, '11': 0.91923809, '6': -1.2751533, '3': -0.61228681, '4': -1.3868549, '2': 0.449507, '1': -0.73226953, '8': 0.50917882, '7': 0.45640573, '0': 0.11466928, '10': -0.23303002, '9': -0.77098644}`

**Comment by AlexKuhnle** [comment]
That's very weird, if I run your script, everything works alright:
```
.../TensorForce$ python test.py
{'0': 0.8950066, '1': -0.048382528, '2': 0.08952066, '3': -1.4426012, '4': 0.3819571, '5': 0.49296466, '6': 0.14841639, '7': 0.5148316, '8': -0.37281656, '9': -0.116880864, '10': 0.92086107, '11': 1.0477791}
.../TensorForce$ python test.py
{'0': 0.8950066, '1': -0.048382528, '2': 0.08952066, '3': -1.4426012, '4': 0.3819571, '5': 0.49296466, '6': 0.14841639, '7': 0.5148316, '8': -0.37281656, '9': -0.116880864, '10': 0.92086107, '11': 1.0477791}
.../TensorForce$ python test.py
{'0': 0.8950066, '1': -0.048382528, '2': 0.08952066, '3': -1.4426012, '4': 0.3819571, '5': 0.49296466, '6': 0.14841639, '7': 0.5148316, '8': -0.37281656, '9': -0.116880864, '10': 0.92086107, '11': 1.0477791}
```
Are you on the latest GitHub version of TensorForce?

**Comment by unrealwill** [comment]
Yes I am on last week github version of TensorForce.
I just recreated a new virtual env to check with a more recent tensorflow version.
```
virtualenv -p python3 venvbug
source ~/venvbug/bin/activate
pip3 install tensorflow-gpu
git clone git@github.com:reinforceio/tensorforce.git
cd tensorforce
pip install -e .
```
It still bug. (It also bug with tensorflow version 1.6)
It's strange. I'll try it on my other machine.

**Comment by AlexKuhnle** [comment]
I'll also try it with a more recent version of TensorFlow soon (used 1.5 so far). If you could also try the CPU version of TensorFlow, that might also be insightful (something seems weird with the more recent GPU versions).

**Comment by unrealwill** [comment]
I tried on an other machine. 

I also tried with tensorflow 1.5
pip3 install tensorflow-gpu==1.5
also tried with : 
pip3 install tensorflow==1.5

In every case it bugs, keys are always in random order.
Very strange indeed.

**Comment by unrealwill** [comment]
I replicated using docker. 
The bug is dependent on python version.

Create a folder. 
Create a file named bugreinforce.py with the code given above
Create a file named Dockerfile with the code given below : 
This dockerfile use python 2.7 and doesn't bug.
```
FROM tensorflow/tensorflow
COPY bugreinforce.py /notebooks/bugreinforce.py
RUN pip install tensorforce
CMD python -v && python /notebooks/bugreinforce.py && python /notebooks/bugreinforce.py 
```

This dockerfile use python3.5 and does bug.
```
FROM tensorflow/tensorflow
COPY bugreinforce.py /notebooks/bugreinforce.py
RUN apt-get update
RUN apt-get -y install python3-pip
RUN pip3 install --upgrade pip
RUN pip3 install tensorforce
RUN pip3 install tensorflow
CMD python3 /notebooks/bugreinforce.py && python3 /notebooks/bugreinforce.py 
```

To build : 
docker build -t bugreinforce .

To run : 
docker run bugreinforce

Output in buggy case : 
```
{'6': 0.33829415, '2': -1.7731481, '3': 0.20157015, '1': 0.23482005, '8': 1.1451101, '9': -0.36537147, '5': -0.1849588, '4': -0.028059486, '11': -0.89746225, '0': 0.6240633, '10': -0.469951, '7': 0.7273865}
{'5': -0.028059486, '4': 0.33829415, '2': 0.7273865, '9': 0.20157015, '0': 0.23482005, '8': 0.6240633, '10': -0.36537147, '1': -0.1849588, '6': 1.1451101, '3': -0.89746225, '7': -1.7731481, '11': -0.469951}
```

**Comment by AlexKuhnle** [comment]
I can reproduce the problem now with Python 3.5 -- interesting that 3.5 behaves so different on dicts. I also have an idea of how to solve it, will let you know.

**Comment by AlexKuhnle** [comment]
This problem should be solved with the latest commit. Please let me know if it is not... :-)

**Comment by unrealwill** [comment]
Hi, I checked your latest commit ( https://github.com/reinforceio/tensorforce/commit/1e7ef7f971dc69d15ecedd47e2dff47317296eda ) , it does fix the bug.

Yet I'm still nervous about the non deterministic .values(), .items() and .keys(), everywhere else in the code.

For example in models/model.py get_variables() : 

```
        if include_nontrainable:
            model_variables = [self.all_variables[key] for key in sorted(self.all_variables)]

            states_preprocessing_variables = [
                variable for preprocessing in self.states_preprocessing.values()
                for variable in preprocessing.get_variables()
            ]
            model_variables += states_preprocessing_variables

            actions_exploration_variables = [
                variable for exploration in self.actions_exploration.values()
                for variable in exploration.get_variables()
            ]
```
The "sorted" (inside the various get_variables()) semantically seems to imply we want deterministic ordering of variables, but the .values() introduce randomness, so it is at best semantically inconsistent, at worst an other silent bug waiting to happen once I introduce back the actions_exploration.

Can you confirm everything is fine with the ~100 references of .items(), .values() and .keys(),... everywhere else in the code.
Thanks

**Comment by AlexKuhnle** [comment]
I went through the occurrences of `.items()` in the various models and 'improved' some of them, as you probably have seen. Why did I pick only these? On the one hand, because I didn't immediately want to spend time going through all the codebase (happy to accept PRs, thought ;-) ). On the other hand, because randomness shouldn't matter in cases where variables/tensors are named or where no TensorFlow operations are generated (and it would crash in the less special cases where there are non-uniform action shapes, I assume). You're right, though, that ideally and to be safe one would just replace all occurrences, maybe this will happen over time.

**Comment by unrealwill** [comment]
Here is an evolved version of the bug : 
afaik turning deterministic = False but setting tensorflow seed and numpy seed should result in pseudo random but repeatable results. (which is the question I was asking when I opened the issue (seeding?) )

The goal is to check that the correct action exploration is applied to the correct action yet (which I don't think it's the case because of get_variables code^^) (they are not used in the code below (set actions_exploration to action_exp instead of None) but you because I already have random order.

```
from tensorforce.agents import PPOAgent
import numpy as np
import tensorflow as tf

np.random.seed(42)
tf.set_random_seed(42)

def demo():
    actions = {}
    actions_exp = {}
    for i in range(12):
        actions[str(i)] = {'type': 'float'}  # 'num_actions': 10
        actions_exp[str(i)] = dict(
            type='ornstein_uhlenbeck',
            sigma=0.1,
            mu=i,
            theta=0.1
        )
    layerSize = 300
    network_spec = [
        dict(type='dense', size=layerSize, activation='tanh'),
        dict(type='dense', size=layerSize, activation='tanh')
    ]

    states = {"jp": dict(type='float', shape=(10,))}

    agent = PPOAgent(
        states=states,
        actions=actions,
        actions_exploration=None,
        batching_capacity=1000,
        network=network_spec,
        step_optimizer=dict(
            type='adam',
            learning_rate=1e-5
        ),
    )

    try :
        agent.restore_model("./bug/")
    except:
        agent.save_model("./bug/")

    _states = {}
    _states["jp"] = np.arange(10)
    print( agent.act( _states, deterministic=False ) )

if __name__ == "__main__":
    demo()
```

This code results in 
First run : 
`{'1': -0.4681145, '11': 3.5902832, '4': -1.5721376, '6': 1.5702705, '5': -1.450424, '3': -1.6934885, '10': 0.61172521, '8': 0.61159497, '0': -0.79051471, '2': 0.31828594, '9': 0.77010918, '7': -7.4641113}`
Second run : 
`{'3': -2.835803, '8': 0.46407062, '9': -0.21255507, '2': 0.97067082, '11': 2.6594412, '6': -1.4957972, '7': -2.0541401, '5': -1.3677678, '0': -0.12639047, '1': 1.3915092, '10': 3.7629828, '4': -1.0760975}`

Thanks

**Comment by Gor-Ren** [comment]
Dicts are guaranteed to maintain insertion order from [Python 3.7 onwards](https://mail.python.org/pipermail/python-dev/2017-December/151283.html), but there is no guarantee in previous Python versions. The underlying dict implementation has changed from time to time (Raymond Hettinger gave a [great talk](https://www.youtube.com/watch?v=p33CVV29OG8) on the subject, if you are interested) such that insertion order is preserved in some Python versions and not in others.

Since TensorForce aims to support a range of Python versions, I'd suggest using [OrderedDict](https://docs.python.org/2/library/collections.html#collections.OrderedDict)s instead, introduced in 2.7, which I think should be a drop-in replacement (from that page: "This allows OrderedDict objects to be substituted anywhere a regular dictionary is used"). This might look nicer than `sorted(...)` everywhere :)

**Comment by AlexKuhnle** [comment]
Should hopefully be fixed now. Let me know if I missed something...
